### ISSUE TITLE
BowPad could crash if both x86 and x64 versions installed with any lang file

### ISSUE THREAD

**Issue by wmjordan** [stefankueng/BowPad#2]
I'd gotten both the 32bit and 64bit version on my machine. They had problem when using localized lang files.

1. Installed the 64bit version and a lang file.
1. Download a 32bit portable version, and fetch the remote lang list.
1. The 32bit version will crash when it uses the same lang as the 64bit version.

Is it possible to place the lang file at the same folder of BowPad (fully portable app), so we can put the x86 and x64 version to different folders and avoid the crash?

**Comment by stefankueng** [comment]
The real problem here is not the language files but the detection of whether an instance of BP is portable or not.
Right now, the detection is done by reading a registry key: if it exists, BP assumes that it was installed, otherwise it assumes it is portable.
So with this detection mechanism, you can't have an installed and a portable version at the same time on the same machine.

Suggestions for other detection mechanisms are welcome...

**Comment by wmjordan** [comment]
The installation was really problematic. If the portable version **check for updates**, after downloading and installing the new version, the portable version becomes an installed version. Thus I suggest you not use the registry detection.

Typical detection of portable version mechanism places a file into the installation directory. Someone uses an empty file named _portable.ini_ (or another name), someone just places the settings file there (when the application finds the settings file there, it assumes that it is running in portable mode).

**Comment by wmjordan** [comment]
Thank you for taking time working with this issue.

I was running BowPad under portable mode (no entry in registry).
While I launched the latest compiled build, it searched for the lang file on the web (I cancelled it since I wanted to use my own lang file) and it removed my localized lang file in the BowPad folder after my cancellation.

Where will the lang file are supposed to be located if BowPad is running in portable mode?

**Comment by stefankueng** [comment]
the lang file is downloaded to the bin folder in portable mode.

**Comment by wmjordan** [comment]
I tried to place the _BowPad_zh-CN.lang_ file into the following positions, but BowPad still insisted to load lang file from the web.

(`.\` stands for the BowPad dir)
````
.\BowPad_zh-CN.lang
.\bin\BowPad_zh-CN.lang
.\BowPad\bin\BowPad_zh-CN.lang
````

**Comment by wmjordan** [comment]
I checked the behavior of BowPad. The lang file was expected to be placed at `.\BowPad_zh-CN.lang` for Chinese language.

My compiled binary BowPad.exe was versioned 2.6.5.0, and the lang file had the same version number.
It seemed that BowPad somehow thought the file was not good and attempted to download it from the web.

**Comment by stefankueng** [comment]
it if had the same version: maybe the reason it failed the check was that the lang file and the BP are not the same bitness (e.g. BP is 64-bit exe and the lang file a 32-bit version or vice versa).

**Comment by wmjordan** [comment]
Both BowPad and the lang file were 64bit.
It had been OK for years before this commit.

**Comment by stefankueng** [comment]
and when you let it download from the web, does it do that repeatedly or does it accept the downloaded version?

**Comment by wmjordan** [comment]
I examined the exe and the lang file and found that they were about 2MB larger than before.
BowPad was 8665KB versus previous version's 6610KB and the lang file was 5292KB vs previous 3240KB.
I tried to rebuild BowPad, got the same result.

When it downloaded the lang file from the web, it did accept the downloaded version.

**Comment by wmjordan** [comment]
I created a branch on my local repository and set it to the last commit before this fix, things were good.

Maybe it is because my localization tool has rebuilt the structure of the lang file and the file does not have the "bitness".

**Comment by wmjordan** [comment]
I debugged the compilation and found the problem.

I've gotten a `.\BowPad_zh-CN.lang` in my BowPad folder, which is actually good. Nothing wrong.

Moreover, in the folder, I also had a `.\BowPad_x86_zh-CN.lang` file, which had been left there since this issue was posted.
At that time, I was attempting to localize BowPad, both x86 and x64 versions. An x86 version of version 2.3 was placed there.

During my debugging, BowPad tried scan its folder for all lang files and checked for their `bitness`es. It went pass the good lang file, and then found the outdated x86 lang file. It therefore somehow treated the good lang file was bad. Then downloaded a version from the web.

I removed the `.\BowPad_x86_zh-CN.lang` file in the directory. BowPad was happy with the `.\BowPad_zh-CN.lang` file then.
